Source/ASDisplayNode.mm

```diff
@@ -1849,30 +1867,31 @@ - (void)calculatedLayoutDidChange
 - (BOOL)displaysAsynchronously
 {
   ASDN::MutexLocker l(__instanceLock__);
-  return [self __locked_displaysAsynchronously];
+  return [self _locked_displaysAsynchronously];
 }
 
 /**
  * Core implementation of -displaysAsynchronously.
- * Must be called with __instanceLock__ held.
  */
-- (BOOL)__locked_displaysAsynchronously
+- (BOOL)_locked_displaysAsynchronously
 {
   return _flags.synchronous == NO && _flags.displaysAsynchronously;
 }
 
 - (void)setDisplaysAsynchronously:(BOOL)displaysAsynchronously
 {
   ASDisplayNodeAssertThreadAffinity(self);
+  
+  ASDN::MutexLocker l(__instanceLock__);
 
   // Can't do this for synchronous nodes (using layers that are not _ASDisplayLayer and so we can't control display prevention/cancel)
-  if (_flags.synchronous)
+  if (_flags.synchronous) {
     return;
+  }
 
-  ASDN::MutexLocker l(__instanceLock__);
-
-  if (_flags.displaysAsynchronously == displaysAsynchronously)
+  if (_flags.displaysAsynchronously == displaysAsynchronously) {
     return;
+  }
 
   _flags.displaysAsynchronously = displaysAsynchronously;
 
```

Comment:
I introduced `_locked_asyncLayer` and was able to almost remove all `asyncLayer` calls as they was  called with the lock held.